### PR TITLE
Robin Leach: Exercise 2

### PR DESCRIPTION
**Pull request by robin-leach**

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -1,25 +1,99 @@
...
+import json
 from flask import Flask, render_template, request, redirect, url_for
-import session_items as session
+import requests
+
+from entity.trello_list import TrelloList 
+from entity.trello_card import TrelloCard
+from entity.item import Item
+
+trello_base_url = os.getenv('TRELLO_API_BASE_URL')
```
Comment (Jonesey13) [inline app.py:10]: Probably worth listing these in the `.env.template` file.

File: app.py [hunk 2]
```diff
@@ -1,25 +1,99 @@
...
 app.config.from_object('flask_config.Config')
 
+def extract_trello_lists(response_item):
+    return TrelloList(response_item['id'], response_item['name'])
+
+def extract_trello_items(response_item):
+    return TrelloCard(response_item['id'], response_item['name'], response_item['idList'])
+
+def get_lists_on_board():
+    endpoint = trello_base_url + '/1/boards/' + trello_board_id + '/lists'
```
Comment (Jonesey13) [inline app.py:25]: Format string syntax would work better here: `f'trello_base_url/1/boards/{trello_board_id}/lists'`.

Also I would argue `/1/` is part of the base url.

File: app.py [hunk 3]
```diff
@@ -1,25 +1,99 @@
...
+    params = { 'key': trello_key, 'token': trello_token }
+    app.logger.debug('Fetching lists from Trello board.\n\tMethod: Get\n\tURL: ' + endpoint + '\n\tParameters: ' + json.dumps(params))
```
Comment (Jonesey13) [inline app.py:27]: A multiline format string would be ideal here: 
```python
f""" Fetching lists from Trello board.
    Method: Get
    URL: {endpoint}
    Parameters: {json.dumps(params)}
"""
```

File: app.py [hunk 4]
```diff
@@ -1,25 +1,99 @@
...
 @app.route('/')
 def index():
-    items = session.get_items()
+    trello_lists = get_lists_on_board()
+    trello_items = get_items_on_board()
+    
+    items = map_trello_items(trello_lists, trello_items)
     return render_template('index.html', items = items)
 
-@app.route('/items/new', methods=['POST'])
```
Comment (Jonesey13) [inline app.py:12]: Removing the ability to add todos is a regression from exercise 1.

File: app.py [hunk 5]
```diff
@@ -1,25 +1,99 @@
...
+from entity.trello_card import TrelloCard
+from entity.item import Item
+
+trello_base_url = os.getenv('TRELLO_API_BASE_URL')
+trello_key = os.getenv('TRELLO_KEY')
+trello_token = os.getenv('TRELLO_TOKEN')
+trello_board_id = os.getenv('TRELLO_BOARD_ID')
 
 app = Flask(__name__)
 app.config.from_object('flask_config.Config')
```
Comment (Jonesey13) [inline app.py:16]: If we're no longer using flask session cookies then the flask_config.py file and SECRET_KEY can be removed

File: templates/index.html [hunk 6]
```diff
@@ -7,25 +7,6 @@ <h1 class="display-4">To-Do App</h1>
     <p class="lead">Just another to-do app.</p>
   </div>
 
-  <div class="row">
-    <div class="col-md-12">
-      <h2>New item</h2>
-      <form action="{{ url_for('add_item') }}" method="POST">
```
Comment (Jonesey13) [inline templates/index.html:13]: Again this is a regression from exercise 1

**Comment by Jonesey13** [comment]
Is there a particular reason we've created a separate repository for this exercise rather than build on https://github.com/robin-leach/DevOps-Course-Exercise-1?

File: app.py (hunk 2 above)
Comment (robin-leach) [reply app.py]: excellent, yes that string formatting is much nicer

**Comment by robin-leach** [comment]
> Is there a particular reason we've created a separate repository for this exercise rather than build on https://github.com/robin-leach/DevOps-Course-Exercise-1?

I sort of assumed we would be creating a new repo for each exercise, but in hindsight that does seem to be quite inefficient

File: .env.template [hunk 7]
```diff
@@ -2,5 +2,8 @@
 FLASK_APP=app
 FLASK_ENV=development
 
-# Change the following values for local development.
-SECRET_KEY=secret-key
+# Trello
+TRELLO_API_BASE_URL=
```
Comment (Jonesey13) [inline .env.template:6]: Might be worth documenting that these should be filled in the README file (along with mentioning the magic list names)

File: helpers/index.py [hunk 8]
```diff
@@ -0,0 +1,23 @@
+import logging
+
+from entity.item import Item
+from trello_requests.trello_requests import get_lists_on_board
+
+log = logging.getLogger('app')
+
+def map_trello_items(trello_lists, trello_items):
+    items = []
+    for trello_item in trello_items:
```
Comment (Jonesey13) [inline helpers/index.py:10]: If you fancy a challenge I bet you could pull this off with one list comprehension.

File: trello_requests/helpers.py [hunk 9]
```diff
@@ -0,0 +1,8 @@
+from entity.trello_card import TrelloCard
+from entity.trello_list import TrelloList
+
+def extract_trello_lists(response_item):
```
Comment (Jonesey13) [inline trello_requests/helpers.py:4]: These 2 could be `@classmethod`s on their corresponding classes

File: entity/http_method.py [hunk 10]
```diff
@@ -0,0 +1,7 @@
+from enum import Enum
+
+class HttpMethod(Enum):
+    Get = "GET"
```
Comment (Jonesey13) [inline entity/http_method.py:4]: Enums options should probably be UPPER_CASE: https://docs.python.org/3/library/enum.html

**Pull request merged**